### PR TITLE
Fix compile error in GCC 8.x series

### PR DESCRIPTION
**Pull request by permcody**
refs #12984

### CODE REVIEW

**Comment by permcody** [comment]
Tag @milljm

**Comment by milljm** [comment]
Additional errors when I run your PR through gcc 8.2.0:

```
/tmp/moose/modules/phase_field/src/postprocessors/FeatureFloodCount.C: In member function 'virtual void FeatureFloodCount::initialize()':
/tmp/moose/modules/phase_field/src/postprocessors/FeatureFloodCount.C:255:45: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
                                             ^
In file included from /tmp/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C:6:
/tmp/moose/modules/phase_field/src/postprocessors/FeatureFloodCount.C: In member function 'void FeatureFloodCount::communicateAndMerge()':
/tmp/moose/modules/phase_field/src/postprocessors/FeatureFloodCount.C:403:38: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
     for (auto i = decltype(_n_vars)(0); i < _n_vars; ++i)
                                      ^
In file included from /tmp/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C:6:
/tmp/moose/modules/phase_field/src/postprocessors/FeatureFloodCount.C: In member function 'virtual void FeatureFloodCount::mergeSets()':
/tmp/moose/modules/phase_field/src/postprocessors/FeatureFloodCount.C:1026:45: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
                                             ^
In file included from /tmp/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C:6:
/tmp/moose/modules/phase_field/src/postprocessors/FeatureFloodCount.C: In member function 'void FeatureFloodCount::consolidateMergedFeatures(std::__debug::vector<std::__debug::list<FeatureFloodCount::FeatureData> >*)':
/tmp/moose/modules/phase_field/src/postprocessors/FeatureFloodCount.C:1094:45: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
                                             ^
In file included from /tmp/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C:8:
/tmp/moose/modules/phase_field/src/postprocessors/GrainBoundaryArea.C: In constructor 'GrainBoundaryArea::GrainBoundaryArea(const InputParameters&)':
/tmp/moose/modules/phase_field/src/postprocessors/GrainBoundaryArea.C:47:43: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   for (auto op_index = decltype(_op_num)(0); op_index < _op_num; ++op_index)
                                           ^
In file included from /tmp/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C:9:
/tmp/moose/modules/phase_field/src/postprocessors/GrainTracker.C: In member function 'void GrainTracker::trackGrains()':
/tmp/moose/modules/phase_field/src/postprocessors/GrainTracker.C:477:49: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
       for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
                                                 ^
In file included from /tmp/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C:9:
/tmp/moose/modules/phase_field/src/postprocessors/GrainTracker.C: In member function 'void GrainTracker::remapGrains()':
/tmp/moose/modules/phase_field/src/postprocessors/GrainTracker.C:964:65: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
           for (auto max = decltype(_max_remap_recursion_depth)(0);
                                                                 ^
/tmp/moose/modules/phase_field/src/postprocessors/GrainTracker.C:1004:67: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
             for (auto max = decltype(_max_remap_recursion_depth)(0);
                                                                   ^
In file included from /tmp/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C:9:
/tmp/moose/modules/phase_field/src/postprocessors/GrainTracker.C: In member function 'virtual void GrainTracker::updateFieldInfo()':
/tmp/moose/modules/phase_field/src/postprocessors/GrainTracker.C:1510:45: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
                                             ^
In file included from /tmp/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C:9:
/tmp/moose/modules/phase_field/src/postprocessors/GrainTracker.C:1565:63: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
           for (auto reserve_index = decltype(_n_reserve_ops)(0); reserve_index < _n_reserve_ops;
                                                               ^
cc1plus: all warnings being treated as errors
/tmp/moose/framework/build.mk:97: recipe for target '/tmp/moose/modules/phase_field/build/unity_src/postprocessors_Unity.x86_64-pc-linux-gnu.dbg.lo' failed
make: *** [/tmp/moose/modules/phase_field/build/unity_src/postprocessors_Unity.x86_64-pc-linux-gnu.dbg.lo] Error 1
make: *** Waiting for unfinished jobs....
In file included from /tmp/moose/modules/phase_field/build/unity_src/userobjects_Unity.C:22:
/tmp/moose/modules/phase_field/src/userobjects/PolycrystalEBSD.C: In member function 'virtual libMesh::Real PolycrystalEBSD::getNodalVariableValue(unsigned int, const libMesh::Node&) const':
/tmp/moose/modules/phase_field/src/userobjects/PolycrystalEBSD.C:76:43: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   for (auto index = decltype(num_grains)(0); index < num_grains; ++index)
                                           ^
cc1plus: all warnings being treated as errors
/tmp/moose/framework/build.mk:97: recipe for target '/tmp/moose/modules/phase_field/build/unity_src/userobjects_Unity.x86_64-pc-linux-gnu.dbg.lo' failed
make: *** [/tmp/moose/modules/phase_field/build/unity_src/userobjects_Unity.x86_64-pc-linux-gnu.dbg.lo] Error 1
In file included from /tmp/moose/modules/phase_field/build/unity_src/materials_Unity.C:5:
/tmp/moose/modules/phase_field/src/materials/ComputePolycrystalElasticityTensor.C: In constructor 'ComputePolycrystalElasticityTensor::ComputePolycrystalElasticityTensor(const InputParameters&)':
/tmp/moose/modules/phase_field/src/materials/ComputePolycrystalElasticityTensor.C:43:43: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   for (auto op_index = decltype(_op_num)(0); op_index < _op_num; ++op_index)
                                           ^
/tmp/moose/modules/phase_field/src/materials/ComputePolycrystalElasticityTensor.C: In member function 'virtual void ComputePolycrystalElasticityTensor::computeQpElasticityTensor()':
/tmp/moose/modules/phase_field/src/materials/ComputePolycrystalElasticityTensor.C:82:43: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   for (auto op_index = decltype(_op_num)(0); op_index < _op_num; ++op_index)
                                           ^
cc1plus: all warnings being treated as errors
/tmp/moose/framework/build.mk:97: recipe for target '/tmp/moose/modules/phase_field/build/unity_src/materials_Unity.x86_64-pc-linux-gnu.dbg.lo' failed
make: *** [/tmp/moose/modules/phase_field/build/unity_src/materials_Unity.x86_64-pc-linux-gnu.dbg.lo] Error 1
```

**Comment by permcody** [comment]
@friedmud - Take a look at this macro and see if you want it in the framework. The problem I'm trying to solve is continuing to support `for (auto ...`, which has gotten more difficult under the GCC 8.x series of compiler (see errors above). We've never been able to use the old macro for POD types but I was able to force something to work by using `decltype` in RHS expressions. The new macro could be used in that new case to solve the compile errors you see above.

I looked into trying to combine the vector and POD types into a single macro but that is a no-go because the conditional statement you end up with also results in different types being generated during expansion for the unused portion of the branch. This can easily be solved with a templated function but we were trying to avoid that for these for-loop I don't think there's a way to solve this with a macro and we were trying to avoid using a templated function for this simple capability.

**Comment by dschwen** [comment]
At this point I have to reraise the question "why not just consistently use `std::size_t`?"

Do we really gain so much from this opaque magic macro? In which cases is size_t incorrect?

**Comment by permcody** [comment]
> At this point I have to reraise the question "why not just consistently use std::size_t? Do we really gain so much from this opaque magic macro? In which cases is size_t incorrect?

That's a fair question and before we decide to go with something like this, it's worth really investigating and definitely answering that question. The immediate example where it would make a difference in MOOSE is in signed vs unsigned cases. There are a few loops that loop to signed quantities and in those cases you have to break your consistent usage and make sure to put in the appropriate signed type.

The other place where it _might_ matter is with vectorization. @friedmud played around with the vectorize flags to the compiler and theorized that mismatched types in the for loops could play into whether or not some loops would vectorize or not. We are interested in trying to take advantage of these vector instructions when we can get the benefit for free. The onus is on us to see if we can indeed verify (re-verify) those old findings before throwing out a much more simple practice of "always using `std::size_t`". In the meantime, this is actually a pretty cool solution for matching types!

**Comment by permcody** [comment]
UPDATE: I only tested with Clang vectorization, but I was unable to produce any cases where the loop counter had an effect one way or the other on loop vectorization. I think the loop vectorization code has come a long way since we last tested. Perhaps there are a few cases that differ with GCC, but I didn't go there. What I will say is that one thing that matters with vectorization is alignment, which may have something to do with types, but I'm not too concerned about that. The whole vectorization argument is actually pretty weak for this.

I still believe that having one consistent interface is reasonable given that we don't live in a world where we can have a single type that works everywhere due to signed-ness. To be clear, I do want to remove the "beginIndex" thing we put in there before and move to this syntax everywhere where indices are needed (at least in the framework). Actually that's not 100% true. I have an "enumerate" function, which is forward compatible with C++17 structured bindings that I'd like to introduce in places where we need values and indices in looping constructs. I'll put up a PR for that later.

**Comment by dschwen** [comment]
How many signed loops do we really have?

**Comment by friedmud** [comment]
This is about always having the correct type.  The correct signedness, 32bit/64bit... having the type change automatically as other parts of MOOSE change (or as you reconfigure, etc.).

This has been a pain to keep right for a long time.  This will make every loop uniform and give us the correct type... a win-win.

As for the number of signed loops - we have quite a few.  PETSc uses only signed integers for instance...

**Comment by permcody** [comment]
> How many signed loops do we really have?

A quick grep tells me there are ~90 in MOOSE + modules. Looks like about ~30 of those are in the framework.

**Review by friedmud — COMMENTED** [review]
👍

**Review by friedmud — APPROVED** [review]
👍

File: framework/include/utils/MooseTypes.h [hunk 1]
```diff
@@ -64,6 +65,39 @@
...
+ * for (MooseIndex(t) i = 0; i < t.size(); ++i)
+ *
+ * POD type:
+ * for (MooseIndex(t) i = 0; i < t; ++i)
+ */
+#define MooseIndex(type) decltype(_MooseIndex(type, 0))
+
+// SFINAE templates for type MooseIndex type selection
+template <typename T, typename std::enable_if<std::is_integral<T>::value>::type * = nullptr>
+typename std::remove_const<T>::type
```
Comment (lindsayad) [inline framework/include/utils/MooseTypes.h]: We might consider also wrapping with `std::remove_reference` although it should be rare that we're ever calling this macro on an integral reference...

**Comment by milljm** [comment]
These warnings persist:
```
Compiling C++ (in dbg mode) /opt/civet/build_0/moose/modules/richards/build/unity_src/dirackernels_Unity.C...
In file included from /opt/civet/build_0/moose/modules/phase_field/build/unity_src/materials_Unity.C:5:
/opt/civet/build_0/moose/modules/phase_field/src/materials/ComputePolycrystalElasticityTensor.C: In constructor 'ComputePolycrystalElasticityTensor::ComputePolycrystalElasticityTensor(const InputParameters&)':
/opt/civet/build_0/moose/modules/phase_field/src/materials/ComputePolycrystalElasticityTensor.C:43:43: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   for (auto op_index = decltype(_op_num)(0); op_index < _op_num; ++op_index)
                                           ^
/opt/civet/build_0/moose/modules/phase_field/src/materials/ComputePolycrystalElasticityTensor.C: In member function 'virtual void ComputePolycrystalElasticityTensor::computeQpElasticityTensor()':
/opt/civet/build_0/moose/modules/phase_field/src/materials/ComputePolycrystalElasticityTensor.C:82:43: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   for (auto op_index = decltype(_op_num)(0); op_index < _op_num; ++op_index)
                                           ^
cc1plus: all warnings being treated as errors
/opt/civet/build_0/moose/framework/build.mk:97: recipe for target '/opt/civet/build_0/moose/modules/phase_field/build/unity_src/materials_Unity.x86_64-pc-linux-gnu.dbg.lo' failed
make: *** [/opt/civet/build_0/moose/modules/phase_field/build/unity_src/materials_Unity.x86_64-pc-linux-gnu.dbg.lo] Error 1
make: *** Waiting for unfinished jobs....
In file included from /opt/civet/build_0/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C:8:
/opt/civet/build_0/moose/modules/phase_field/src/postprocessors/GrainBoundaryArea.C: In constructor 'GrainBoundaryArea::GrainBoundaryArea(const InputParameters&)':
/opt/civet/build_0/moose/modules/phase_field/src/postprocessors/GrainBoundaryArea.C:47:43: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   for (auto op_index = decltype(_op_num)(0); op_index < _op_num; ++op_index)
                                           ^
In file included from /opt/civet/build_0/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C:9:
/opt/civet/build_0/moose/modules/phase_field/src/postprocessors/GrainTracker.C: In member function 'void GrainTracker::trackGrains()':
/opt/civet/build_0/moose/modules/phase_field/src/postprocessors/GrainTracker.C:477:49: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
       for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
                                                 ^
In file included from /opt/civet/build_0/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C:9:
/opt/civet/build_0/moose/modules/phase_field/src/postprocessors/GrainTracker.C: In member function 'void GrainTracker::remapGrains()':
/opt/civet/build_0/moose/modules/phase_field/src/postprocessors/GrainTracker.C:964:65: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
           for (auto max = decltype(_max_remap_recursion_depth)(0);
                                                                 ^
/opt/civet/build_0/moose/modules/phase_field/src/postprocessors/GrainTracker.C:1004:67: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
             for (auto max = decltype(_max_remap_recursion_depth)(0);
                                                                   ^
In file included from /opt/civet/build_0/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C:9:
/opt/civet/build_0/moose/modules/phase_field/src/postprocessors/GrainTracker.C: In member function 'virtual void GrainTracker::updateFieldInfo()':
/opt/civet/build_0/moose/modules/phase_field/src/postprocessors/GrainTracker.C:1510:45: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
                                             ^
In file included from /opt/civet/build_0/moose/modules/phase_field/build/unity_src/postprocessors_Unity.C:9:
/opt/civet/build_0/moose/modules/phase_field/src/postprocessors/GrainTracker.C:1565:63: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
           for (auto reserve_index = decltype(_n_reserve_ops)(0); reserve_index < _n_reserve_ops;
                                                               ^
cc1plus: all warnings being treated as errors
/opt/civet/build_0/moose/framework/build.mk:97: recipe for target '/opt/civet/build_0/moose/modules/phase_field/build/unity_src/postprocessors_Unity.x86_64-pc-linux-gnu.dbg.lo' failed
make: *** [/opt/civet/build_0/moose/modules/phase_field/build/unity_src/postprocessors_Unity.x86_64-pc-linux-gnu.dbg.lo] Error 1
In file included from /opt/civet/build_0/moose/modules/phase_field/build/unity_src/userobjects_Unity.C:22:
/opt/civet/build_0/moose/modules/phase_field/src/userobjects/PolycrystalEBSD.C: In member function 'virtual libMesh::Real PolycrystalEBSD::getNodalVariableValue(unsigned int, const libMesh::Node&) const':
/opt/civet/build_0/moose/modules/phase_field/src/userobjects/PolycrystalEBSD.C:76:43: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
   for (auto index = decltype(num_grains)(0); index < num_grains; ++index)
                                           ^
cc1plus: all warnings being treated as errors
/opt/civet/build_0/moose/framework/build.mk:97: recipe for target '/opt/civet/build_0/moose/modules/phase_field/build/unity_src/userobjects_Unity.x86_64-pc-linux-gnu.dbg.lo' failed
make: *** [/opt/civet/build_0/moose/modules/phase_field/build/unity_src/userobjects_Unity.x86_64-pc-linux-gnu.dbg.lo] Error 1
ERROR: exiting with code 2
```

**Comment by permcody** [comment]
Are you sure you are up to date?

**Pull request merged**